src/cmd/services/m3coordinator/downsample/metrics_appender.go

```diff
@@ -363,13 +399,184 @@ func (a *metricsAppender) debugLogMatch(str string, opts debugLogMatchOptions) {
 func (a *metricsAppender) NextMetric() {
 	a.tags.names = a.tags.names[:0]
 	a.tags.values = a.tags.values[:0]
+
+	// Move the inuse encoders to cached as we should be done with using them.
+	a.moveInuseEncoders()
 }
 
 func (a *metricsAppender) Finalize() {
 	// Return to pool.
 	a.pool.Put(a)
 }
 
+func (a *metricsAppender) moveInuseEncoders() {
+	a.cachedEncoders = append(a.cachedEncoders, a.inuseEncoders...)
+	for i := range a.inuseEncoders {
+		a.inuseEncoders[i] = nil
+	}
+	a.inuseEncoders = a.inuseEncoders[:0]
+}
+
+func (a *metricsAppender) tagEncoder() serialize.TagEncoder {
+	// Take an encoder from the cached encoder list, if not present get one
+	// from the pool. Add the returned encoder to the used list.
+	var tagEncoder serialize.TagEncoder
+	if len(a.cachedEncoders) == 0 {
+		tagEncoder = a.tagEncoderPool.Get()
+	} else {
+		tagEncoder = a.cachedEncoders[0]
+		a.cachedEncoders = a.cachedEncoders[1:]
+	}
+	a.inuseEncoders = append(a.inuseEncoders, tagEncoder)
+	tagEncoder.Reset()
+	return tagEncoder
+}
+
+func (a *metricsAppender) newSamplesAppenders(
+	stagedMetadata metadata.StagedMetadata,
+	unownedID []byte,
+) ([]samplesAppender, error) {
```

Comment:
This code didn't allocate before, it used to call `a.multiSamplesAppender.addSamplesAppender(samplesAppender{...})` which used a cached slice under the hood in the multiSamplesAppender. Can we make it do the same so we're not allocating a slice here to return something?

Reply:
Can we just make this not return anything and simply call it "addSamplesAppenders" or something and have it call "a.multiSamplesAppender.addSamplesAppender(...)" internally instead of return `[]samplesAppender` which then directly gets added to the multiSamplesAppender anyhow?

Reply:
yeah, can just make that method add to the multi samples appender.. will call the method `addSamplesAppenders`

Reply:
didnt see the comment above, but yeah did that.